### PR TITLE
Atx to include the proof of the initialization phase

### PR DESCRIPTION
**Pull request by moshababo**
Closing https://github.com/spacemeshos/go-spacemesh/issues/1075.

* `commitment` field was added to atx
* `commitmentMerkleRoot` was added to nipst challenge (instead of adding `commitment` hash)
* atx syntactic validation to verify these fields
* atx builder to support an already-initialized state, and to execute a zero-challenge proof on the existing files

### CODE REVIEW

File: activation/activation.go [hunk 1]
```diff
@@ -244,34 +249,47 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
...
+		b.log.Info("PoST initialization completed")
+		atomic.StoreInt32(&b.initStatus, InitDone)
 	}()
-	b.postInitLock.Unlock()
+
 	// todo: refactor initializePost to be concurrent, returning error in a blocking manner
-	t := time.NewTimer(b.errorTimeoutMs * time.Millisecond)
 	select {
-	case <-t.C:
+	case <-time.After(100 * time.Millisecond):
```
Comment (y0sher) [inline activation/activation.go]: move out to constant/config

File: activation/activation.go [hunk 2]
```diff
@@ -421,10 +444,9 @@ func (b *Builder) PublishActivationTx(epoch types.EpochId) error {
 		log.Error("cannot discard nipst challenge %s", err)
 	}
 
-	b.log.Event().Info("atx published!", log.AtxId(atx.ShortId()), log.String("prev_atx_id", atx.PrevATXId.ShortString()),
-		log.String("post_atx_id", atx.PositioningAtx.ShortString()), log.LayerId(uint64(atx.PubLayerIdx)), log.EpochId(uint64(atx.PubLayerIdx.GetEpoch(b.layersPerEpoch))),
-		log.Uint32("active_set", atx.ActiveSetSize), log.String("miner", b.nodeId.Key[:5]), log.Int("view", len(atx.View)))
-
+	b.log.Event().Info(fmt.Sprintf("atx published! id: %v, prevATXID: %v, posATXID: %v, layer: %v, published in epoch: %v, active set: %v miner: %v view %v",
```
Comment (y0sher) [inline activation/activation.go]: why did you change it back to old format ? checkout develop or rebase.

File: activation/atxdb.go [hunk 3]
```diff
@@ -215,9 +216,30 @@ func (db *ActivationDb) SyntacticallyValidateAtx(atx *types.ActivationTx) error
...
 			return fmt.Errorf("sequence number is not one more than prev sequence number")
 		}
+
+		if atx.Commitment != nil {
+			return fmt.Errorf("prevATX declared, but commitment proof is included")
+		}
+
+		if atx.CommitmentMerkleRoot != nil {
+			return fmt.Errorf("prevATX declared, but commitment merkle root is included in challenge")
+		}
```
Comment (y0sher) [inline activation/atxdb.go]: you could just return here instead of `else`

File: activation/activation.go [hunk 4]
```diff
@@ -244,34 +249,47 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
...
-	}
-	if b.initStatus == InProgress {
-		b.postInitLock.Unlock()
-		return fmt.Errorf("attempted to start post when post already initiated")
+
+	initStatus := atomic.LoadInt32(&b.initStatus)
+
+	switch initStatus {
+	case InitDone:
+		return fmt.Errorf("already initialized")
```
Comment (antonlerner) [inline activation/activation.go]: Why should we return an error here? why not success... we've initialized the post
also, does this cover the case where start post is called with different parameters? post will need to re run

File: activation/activation.go [hunk 5]
```diff
@@ -244,34 +249,47 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
...
 	b.SetCoinbaseAccount(rewardAddress)
-	b.postInitLock.Lock()
-	if b.initStatus == Done {
-		return nil
-	}
-	if b.initStatus == InProgress {
-		b.postInitLock.Unlock()
-		return fmt.Errorf("attempted to start post when post already initiated")
+
+	initStatus := atomic.LoadInt32(&b.initStatus)
```
Comment (antonlerner) [inline activation/activation.go]: this does not guarantee we will enter start post only once.

File: activation/activation.go (hunk 1 above)
Comment (moshababo) [reply activation/activation.go]: I removed it from the atx builder struct on purpose, so we won't get used to it. The timeout solution is bad and should be removed.

File: activation/activation.go (hunk 5 above)
Comment (moshababo) [reply activation/activation.go]: Do you refer to the lines before that, or in general?

File: activation/activation.go (hunk 4 above)
Comment (moshababo) [reply activation/activation.go]: > Why should we return an error here? why not success... we've initialized the post

`StartPost` is async, and supposed to start something. If it's already initialized, then
nothing is going to start, and the user should know about this (at least in the api level).
Also, because it's async we're not returning successes, only validation errors. So even if it was a success, it would have created a clumsy api. In this case, the solution would be to swallow the error, to return nothing, and so to have less feedback to the user.

> also, does this cover the case where start post is called with different parameters? post will need to re run

`datadir` is being set from the user input. 
`space`- if there are no init files, the user `space` will apply. if there are, then he will get a `config mismatch` error. IMO we should not delete the files for the user implicitly. He should do so with a different api call.

File: nipst/post.go [hunk 6]
```diff
@@ -21,78 +19,74 @@ func DefaultConfig() config.Config {
...
-func (c *PostClient) SetLogger(logger shared.Logger) {
-	c.logger = logger
+func (c *PostClient) IsInitialized() bool {
+	state, _, _ := c.initializer.State()
+	return state == initialization.StateCompleted
 }
 
 func (c *PostClient) SetParams(dataDir string, space uint64) {
 	cfg := *c.cfg
 	c.cfg = &cfg
```
Comment (antonlerner) [inline nipst/post.go:104]: perhaps do this last

File: nipst/post.go [hunk 7]
```diff
@@ -21,78 +19,74 @@ func DefaultConfig() config.Config {
...
-	if c.initializer == nil {
-		return false
-	}
+func (c *PostClient) SetLogger(logger shared.Logger) {
+	c.logger = logger
+}
 
-	state, _, _ := c.initializer.State()
-	return state == initialization.StateCompleted
+func (c *PostClient) Cfg() *config.Config {
```
Comment (antonlerner) [inline nipst/post.go:123]: To what purpose we have this accessor? is it supposed to allow modifications? If so, ,ultithreading issue should be taken into account, if not, I'd recommend returning a copy

File: nipst/post.go (hunk 6 above)
Comment (moshababo) [reply nipst/post.go:104]: I thought so, but I followed what you did. Will change.

File: nipst/post.go (hunk 7 above)
Comment (moshababo) [reply nipst/post.go:123]: The purpose is to remove `postCfg` from from `nipstBuilder`, so we won't have it in 2 places.
In this case, it need to be read from `postClient`. 

So this isn't to allow writes, only reads.
I'll return a copy instead, as you suggested.

File: nipst/post.go (hunk 6 above)
Comment (moshababo) [reply nipst/post.go:104]: Fixed.

File: nipst/post.go (hunk 7 above)
Comment (moshababo) [reply nipst/post.go:123]: Fixed.

**Comment by moshababo** [comment]
@y0sher @antonlerner I removed `StartPost` timeout workaround, see new implementation.

File: activation/activation.go (hunk 5 above)
Comment (moshababo) [reply activation/activation.go]: See new implementation.

File: activation/activation.go [hunk 8]
```diff
@@ -92,15 +93,13 @@ type Builder struct {
 	store           BytesStore
 	isSynced        func() bool
 	accountLock     sync.RWMutex
-	postInitLock    sync.RWMutex
+	InitLock        sync.RWMutex
```
Comment (zalmen) [inline activation/activation.go]: it is only used internally, change to `initLock`

Comment (zalmen) [reply activation/activation.go]: Also, currently this lock has 3 purposes -
1. Protect initStatus
2. Make sure we execute `StartPost` only once
3. Protect the `dataDir` value within `PostProverClient`

IMO it is an abuse and these three scenarios have nothing in common. 1 should be solved with atomic value, 2 with a mutex and 3 inside the post prover

File: activation/activation.go [hunk 9]
```diff
@@ -197,7 +197,11 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
 	}
 	seq := uint64(0)
 	prevAtxId := *types.EmptyAtxId
-	if b.prevATX != nil {
+	commitmentMerkleRoot := []byte(nil)
+
+	if b.prevATX == nil {
+		commitmentMerkleRoot = b.commitment.MerkleRoot
```
Comment (zalmen) [inline activation/activation.go:200]: might be good to add a comment explaining the purpose of this comm. and why is it only set in this case

File: activation/activation.go (hunk 4 above)
Comment (zalmen) [reply activation/activation.go]: Few things -
> Why should we return an error here? why not success

Personally, I agree with @antonlerner. If the client sends `StartMining` request and gets an error, he might assume that there is no mining going on. However, I talked about it with @IlyaVi, and he pointed out that before calling `StartMining` the client asks for the mining status, which should indicate a running init, and therefore it is indeed an error to re-run the mining.
> post will need to re run

I disagree. At this point (MVP), once an init had started it should never get re-run again (not even after completion). Later we'll support replacing PoET servers, extending/reducing post space and relocating init files.

File: activation/activation.go [hunk 10]
```diff
@@ -243,50 +250,82 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
...
+	switch b.initStatus {
+	case InitDone:
+		b.InitLock.Unlock()
+		return fmt.Errorf("already initialized")
+	case InitInProgress:
+		b.InitLock.Unlock()
+		return fmt.Errorf("already started")
+	}
+
 	b.log.Info("Starting post, reward address: %x", rewardAddress)
```
Comment (zalmen) [inline activation/activation.go]: consider printing also the data dir and space, will ease debugging

File: activation/activation.go [hunk 11]
```diff
@@ -243,50 +250,82 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
...
+	b.postProver.SetParams(dataDir, space)
+	b.InitLock.Unlock()
+
 	b.SetCoinbaseAccount(rewardAddress)
-	b.postInitLock.Lock()
-	if b.initStatus == Done {
-		b.postInitLock.Unlock()
-		return nil
+
+	initialized, err := b.postProver.IsInitialized()
```
Comment (zalmen) [inline activation/activation.go:263]: Why do you need to wait for "the last minute" to check this? Why won't you check it when the node starts (and before the rpc is open for service)?

File: activation/activation.go [hunk 12]
```diff
@@ -243,50 +250,82 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
...
+			b.commitment, err = b.postProver.Initialize()
+			if err != nil {
+				b.log.Error("PoST initialization failed: %v", err)
+
+				b.InitLock.Lock()
+				b.initStatus = InitIdle
+				b.InitLock.Unlock()
+				return
+			}
 		}
```
Comment (zalmen) [inline activation/activation.go:301]: What about running the initial post in the case that `initialized` is `false`?

File: activation/activation.go [hunk 13]
```diff
@@ -243,50 +250,82 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
...
+			if err != nil {
+				b.log.Error("PoST execution failed: %v", err)
+
+				b.InitLock.Lock()
+				b.initStatus = InitIdle
+				b.InitLock.Unlock()
+				return
+			}
+		} else {
+			b.commitment, err = b.postProver.Initialize()
```
Comment (zalmen) [inline activation/activation.go:295]: we have another problem (outside the scope of this PR) that we can't stop the init process when doing graceful shutdown

File: activation/activation.go [hunk 14]
```diff
@@ -197,7 +197,11 @@ func (b *Builder) buildNipstChallenge(epoch types.EpochId) error {
 	}
 	seq := uint64(0)
 	prevAtxId := *types.EmptyAtxId
-	if b.prevATX != nil {
+	commitmentMerkleRoot := []byte(nil)
+
+	if b.prevATX == nil {
```
Comment (barakshani) [inline activation/activation.go:196]: you already check this condition a few lines above

**Review by barakshani — APPROVED** [review]

File: activation/activation.go (hunk 14 above)
Comment (moshababo) [reply activation/activation.go:196]: The previous section is about fetching `prevATX` if exists, and this one is about handling different stuff whenever it exist or not. 
I thought putting the new logic in the previous section would be confusing and easy to miss.

File: nipst/nipst.go [hunk 15]
```diff
@@ -197,18 +196,14 @@ func (nb *NIPSTBuilder) BuildNIPST(challenge *common.Hash) (*types.NIPST, error)
...
 
-		proof, err := nb.postProver.execute(nb.id, nb.state.PoetProofRef, defTimeout)
+		proof, err := nb.postProver.Execute(nb.state.PoetProofRef)
 		if err != nil {
 			return nil, fmt.Errorf("failed to execute PoST: %v", err)
 		}
 
-		res, err := nb.verifyPost(proof, nb.postCfg.SpacePerUnit, nb.postCfg.NumProvenLabels, nb.postCfg.Difficulty)
-		if err != nil {
+		if err := nb.verifyPost(proof, cfg.SpacePerUnit, cfg.NumProvenLabels, cfg.Difficulty); err != nil {
```
Comment (zalmen) [inline nipst/nipst.go]: Do you see any reason to keep this defensive behavior?

File: nipst/post.go [hunk 16]
```diff
@@ -3,96 +3,130 @@ package nipst
...
 }
 
-func (c *PostClient) initialize(id []byte, timeout time.Duration) (commitment *types.PostProof, err error) {
-	// TODO(moshababo): implement timeout
-	//TODO: implement persistence
-	if c.cfg.SpacePerUnit%merkle.NodeSize != 0 {
-		return nil, fmt.Errorf("space (%d) is not a multiple of merkle.NodeSize (%d)", c.cfg.SpacePerUnit, merkle.NodeSize)
-	}
-	c.initializer = initialization.NewInitializer(c.cfg, id)
+func (c *PostClient) Initialize() (commitment *types.PostProof, err error) {
```
Comment (zalmen) [inline nipst/post.go:55]: `Initialize` is a very long operation, are you sure that you have to hold the lock for the entire duration? For example, if, during initialization, the wallet will call `StartPost` it will be delayed until the previous `Initialize` call returns.

**Review by zalmen — CHANGES_REQUESTED** [review]

File: activation/activation.go (hunk 4 above)
Comment (moshababo) [reply activation/activation.go]: > Personally, I agree with @antonlerner. If the client sends StartMining request and gets an error, he might assume that there is no mining going on. However, I talked about it with @IlyaVi, and he pointed out that before calling StartMining the client asks for the mining status, which should indicate a running init, and therefore it is indeed an error to re-run the mining.

He wouldn’t assume that there’s no mining going if he’ll get the proper error msg.

File: activation/activation.go (hunk 8 above)
Comment (moshababo) [reply activation/activation.go]: I assume this comment was on outdated code. See latest changes.

File: activation/activation.go (hunk 9 above)
Comment (moshababo) [reply activation/activation.go:200]: I added this:
```
// if and only if it's the first ATX, the merkle root of the initial PoST proof,
// the commitment, is included in the nipst challenge. This is done in order to prove
// that it existed before the PoET start time.
```

File: activation/activation.go (hunk 10 above)
Comment (moshababo) [reply activation/activation.go]: Done.

File: activation/activation.go (hunk 12 above)
Comment (moshababo) [reply activation/activation.go:301]: I'm not sure on what code did you wrote that comment.
But the initial post _is_ being run in case that `initialized` is false.
The only case we're not doing that is when `VerifyInitAllowed` returns an err.
We're doing this check before launching the goroutine, to return an immediate err to the client.

File: activation/activation.go (hunk 13 above)
Comment (moshababo) [reply activation/activation.go:295]: Indeed - we would either need to add native support for that in `spacemeshos/post`, or use its server mode, and kill its process at will. The later shouldn't be too complicated, due to `post/integration` package. It's already done in unit tests.

File: nipst/nipst.go (hunk 15 above)
Comment (moshababo) [reply nipst/nipst.go]: Probably not at this point. Removed.

File: nipst/post.go (hunk 16 above)
Comment (moshababo) [reply nipst/post.go:55]: Good point, although that shouldn't happen, because the atx builder state protects against multiple calls.
I should anyway reduce  the lock to protect only the struct state, as originally intended. To do so, I can change it to from `Lock` to `RLock`. 
But `SetParams` must remain `Lock`, and so the problem you described can still occur, if not protected in a higher scope. 
I can use atomic ops and variable state instead, and return errors instead of introducing waits, but it looks like an overkill to me.

Let me know what you think.

File: activation/activation.go (hunk 11 above)
Comment (moshababo) [reply activation/activation.go:263]: We don't, but a "last minute" check is the most accurate and simple (no need to manage state), and considering that this is a very quick check, I didn't thought to do it prior to the call.

Should I change that?

File: activation/activation.go (hunk 10 above)
Comment (zalmen) [reply activation/activation.go]: Please use `With()`

File: activation/activation.go (hunk 11 above)
Comment (zalmen) [reply activation/activation.go:263]: if it is quick then it doesn't matter much (when writing this comment I thought that we are actually validating the data on disk which might take some time), leave it as is

File: nipst/nipst.go [hunk 17]
```diff
@@ -197,19 +192,11 @@ func (nb *NIPSTBuilder) BuildNIPST(challenge *common.Hash) (*types.NIPST, error)
...
 
-		res, err := nb.verifyPost(proof, nb.postCfg.SpacePerUnit, nb.postCfg.NumProvenLabels, nb.postCfg.Difficulty)
-		if err != nil {
-			return nil, fmt.Errorf("created an invalid PoST proof: %v", err)
-		}
-		if !res {
-			return nil, fmt.Errorf("created an invalid PoST proof")
-		}
-
 		nb.log.Info("finished PoST execution (proof: %v)", proof)
```
Comment (zalmen) [inline nipst/nipst.go]: not sure if it is a good idea to print the entire post proof..

File: activation/activation.go (hunk 10 above)
Comment (moshababo) [reply activation/activation.go]: Done.

File: nipst/nipst.go (hunk 17 above)
Comment (moshababo) [reply nipst/nipst.go]: Fixed.

**Review by zalmen — APPROVED** [review]
